### ISSUE TITLE
Should have better IDEs support

### ISSUE THREAD

**Issue by ralfbiedert** [rustwasm/wasm-bindgen#2214]
### Motivation

I have the same problem as @panstromek in [this PR](https://github.com/rustwasm/wasm-bindgen/pull/2012). With the changes that landed there code completion in IntelliJ broke; and in turn yew + web-sys became unusable for me. 

One could argue this is an IDE issue and having them fix the problem would be nice, but for example the [underlying IntelliJ ticket has been around for 3 years](https://github.com/intellij-rust/intellij-rust/issues/1786), so it's unlikely to happen soon. 

For most other libs I'd just switch to something with autocomplete support, but `web-sys` has a quasi monopoly on client-web interop, so one could also argue there is some shared responsibility. 

### Possible Solution?

Microsoft's [win-rt](https://github.com/microsoft/winrt-rs) uses `build.rs` + proc macros to create a `winrt.rs`, [as described here](https://kennykerr.ca/2020/06/09/improving-the-ide-for-rust-winrt/), which can be included via `include!(concat!(env!("OUT_DIR"), "/winrt.rs"));`. This works really well for me. I am wondering if something similar is possible here?

**Comment by alexcrichton** [comment]
Thanks for the report! I'm happy to review PRs and/or discuss possible ways to solve this, but unfortunately I don't have time to work on this myself. That PR is beneficial enough that I do not think it's worth reverting, additionally.

**Comment by vlad20012** [comment]
Do I understand correctly that previously all wasm-bindgen work was done in during `build.rs` step yielding complete IDE-supported code, but now most of the wasm-bindgen work is done by procedural macros?

**Comment by panstromek** [comment]
@vlad20012 no, this was true only for `web-sys` crate. All wasm-bindgen stuff is done through proc-macros, just the `web-sys` crate used to be 
an exception generated with a build script (I guess for historical reasons). That mentioned PR just unified it with `js-sys` and all other user code that uses wasm-bindgen.

**Comment by panstromek** [comment]
@vlad20012 now, when you're here, I thought about this some more and I think it shouldn't be that complicated to support it in the plugin, if you're willing to special case it, because `wasm-bindgen` uses pretty predictable structure and the support is "almost there". Plugin can already resolve the functions and types but namespaces them incorrectly.

Basically this is the failing structure:
```rust
#[wasm_bindgen]
extern "C" {
  type SomeJsType;
  fn method(this: &SomeJsType);
}
```
Plugin sees this as a function and unrelated type but it should interpret it as:
```rust
struct SomeJsType;
impl SomeJsType {
  fn method(&self){}
}
```
This would make the resolution correct.

**Comment by vlad20012** [comment]
Oh, let's just support proc macros

**Comment by panstromek** [comment]
Well.. if you think that's simpler solution 😄 😄
